File: db4on/core/src/core/com/db4o/internal/PersistentBase.cs
namespace com.db4o.@internal
{
	/// <exclude></exclude>
	public abstract class PersistentBase : com.db4o.@internal.Persistent
	{
		protected int i_id;

		protected int i_state = 2;

		internal bool BeginProcessing()
		{
			if (BitIsTrue(com.db4o.@internal.Const4.PROCESSING))
			{
				return false;
			}
			BitTrue(com.db4o.@internal.Const4.PROCESSING);
			return true;
		}

		internal void BitFalse(int bitPos)
		{
			i_state &= ~(1 << bitPos);
		}

		internal bool BitIsFalse(int bitPos)
		{
			return (i_state | (1 << bitPos)) != i_state;
		}

		internal bool BitIsTrue(int bitPos)
		{
			return (i_state | (1 << bitPos)) == i_state;
		}

		internal void BitTrue(int bitPos)
		{
			i_state |= (1 << bitPos);
		}

		internal virtual void CacheDirty(com.db4o.foundation.Collection4 col)
		{
			if (!BitIsTrue(com.db4o.@internal.Const4.CACHED_DIRTY))
			{
				BitTrue(com.db4o.@internal.Const4.CACHED_DIRTY);
				col.Add(this);
			}
		}

		public virtual void EndProcessing()
		{
			BitFalse(com.db4o.@internal.Const4.PROCESSING);
		}

		public virtual int GetID()
		{
			return i_id;
		}

		public bool IsActive()
		{
			return BitIsTrue(com.db4o.@internal.Const4.ACTIVE);
		}

		public virtual bool IsDirty()
		{
			return BitIsTrue(com.db4o.@internal.Const4.ACTIVE) && (!BitIsTrue(com.db4o.@internal.Const4
				.CLEAN));
		}

		public bool IsNew()
		{
			return i_id == 0;
		}

		public virtual int LinkLength()
		{
			return com.db4o.@internal.Const4.ID_LENGTH;
		}

		internal void NotCachedDirty()
		{
			BitFalse(com.db4o.@internal.Const4.CACHED_DIRTY);
		}

		public virtual void Read(com.db4o.@internal.Transaction trans)
		{
			try
			{
				if (BeginProcessing())
				{
					com.db4o.@internal.Buffer reader = trans.Stream().ReadReaderByID(trans, GetID());
					if (reader != null)
					{
						ReadThis(trans, reader);
						SetStateOnRead(reader);
					}
					EndProcessing();
				}
			}
			catch (System.Exception t)
			{
			}
		}

		public virtual void SetID(int a_id)
		{
			i_id = a_id;
		}

		public void SetStateClean()
		{
			BitTrue(com.db4o.@internal.Const4.ACTIVE);
			BitTrue(com.db4o.@internal.Const4.CLEAN);
		}

		public void SetStateDeactivated()
		{
			BitFalse(com.db4o.@internal.Const4.ACTIVE);
		}

		public virtual void SetStateDirty()
		{
			BitTrue(com.db4o.@internal.Const4.ACTIVE);
			BitFalse(com.db4o.@internal.Const4.CLEAN);
		}

		internal virtual void SetStateOnRead(com.db4o.@internal.Buffer reader)
		{
			if (BitIsTrue(com.db4o.@internal.Const4.CACHED_DIRTY))
			{
				SetStateDirty();
			}
			else
			{
				SetStateClean();
			}
		}

		public void Write(com.db4o.@internal.Transaction trans)
		{
			if (!WriteObjectBegin())
			{
				return;
			}
			com.db4o.@internal.LocalObjectContainer stream = (com.db4o.@internal.LocalObjectContainer
				)trans.Stream();
			int address = 0;
			int length = OwnLength();
			com.db4o.@internal.Buffer writer = new com.db4o.@internal.Buffer(length);
			if (IsNew())
			{
				com.db4o.@internal.slots.Pointer4 ptr = stream.NewSlot(trans, length);
				SetID(ptr._id);
				address = ptr._address;
			}
			else
			{
				address = stream.GetSlot(length);
				trans.SlotFreeOnRollbackCommitSetPointer(i_id, address, length);
			}
			WriteThis(trans, writer);
			writer.WriteEncrypt(stream, address, 0);
			if (IsActive())
			{
				SetStateClean();
			}
			EndProcessing();
		}

		public virtual bool WriteObjectBegin()
		{
			if (IsDirty())
			{
				return BeginProcessing();
			}
			return false;
		}

		public virtual void WriteOwnID(com.db4o.@internal.Transaction trans, com.db4o.@internal.Buffer
			 writer)
		{
			Write(trans);
			writer.WriteInt(GetID());
		}

		public abstract byte GetIdentifier();

		public abstract int OwnLength();

		public abstract void ReadThis(com.db4o.@internal.Transaction arg1, com.db4o.@internal.Buffer
			 arg2);

		public abstract void WriteThis(com.db4o.@internal.Transaction arg1, com.db4o.@internal.Buffer
			 arg2);
	}
}

File: db4oj/core/src/com/db4o/internal/PersistentBase.java
/* Copyright (C) 2004   Versant Inc.   http://www.db4o.com */

package com.db4o.internal;

import com.db4o.*;
import com.db4o.foundation.*;
import com.db4o.internal.slots.*;


/**
 * @exclude
 */
public abstract class PersistentBase extends Identifiable implements Persistent, LinkLengthAware {
	
	
	public PersistentBase(){
		
	}

    void cacheDirty(Collection4 col) {
        if (!bitIsTrue(Const4.CACHED_DIRTY)) {
            bitTrue(Const4.CACHED_DIRTY);
            col.add(this);
        }
    }

    public void free(LocalTransaction trans){
    	trans.systemTransaction().idSystem().notifySlotDeleted(getID(), slotChangeFactory());
    }

    public final int linkLength() {
        return Const4.ID_LENGTH;
    }

    final void notCachedDirty() {
        bitFalse(Const4.CACHED_DIRTY);
    }

    public void read(Transaction trans) {
		if (!beginProcessing()) {
			return;
		}
		try {
			ByteArrayBuffer reader = produceReadBuffer(trans); 
			
			if (Deploy.debug) {
				reader.readBegin(getIdentifier());
			}
			readThis(trans, reader);
			setStateOnRead(reader);
		} finally {
			endProcessing();
		}
	}
    
    protected ByteArrayBuffer produceReadBuffer(Transaction trans){
    	return readBufferById(trans);
    }
    
    protected ByteArrayBuffer readBufferById(Transaction trans){
    	return trans.container().readBufferById(trans, getID());
    }
    
    void setStateOnRead(ByteArrayBuffer reader) {
        if (Deploy.debug) {
            reader.readEnd();
        }
        if (bitIsTrue(Const4.CACHED_DIRTY)) {
            setStateDirty();
        } else {
            setStateClean();
        }
    }

    public void write(Transaction trans) {
        if (! writeObjectBegin()) {
            return;
        }
        try {
	            
	        LocalObjectContainer container = (LocalObjectContainer)trans.container();
	        
	        if(DTrace.enabled){
	            DTrace.PERSISTENT_OWN_LENGTH.log(getID());
	        }
	        
	        int length = ownLength();
	        length = container.blockConverter().blockAlignedBytes(length);
	        
	        Slot slot = container.allocateSlot(length);
	        
	        if(isNew()){
	            setID(trans.idSystem().newId(slotChangeFactory()));
                trans.idSystem().notifySlotCreated(_id, slot, slotChangeFactory());
	        }else{
	            trans.idSystem().notifySlotUpdated(_id, slot, slotChangeFactory());
	        }
	        
	        ByteArrayBuffer writer = produceWriteBuffer(trans, length);
	        
	        writeToFile(trans, writer, slot);
        }finally{
        	endProcessing();
        }

    }

	protected ByteArrayBuffer produceWriteBuffer(Transaction trans, int length) {
		return newWriteBuffer(length);
	}
	
	protected ByteArrayBuffer newWriteBuffer(int length) {
		return new ByteArrayBuffer(length);
	}
    
	private final void writeToFile(Transaction trans, ByteArrayBuffer writer, Slot slot) {
		
        if(DTrace.enabled){
        	DTrace.PERSISTENTBASE_WRITE.log(getID());
        }
		
		LocalObjectContainer container = (LocalObjectContainer)trans.container();
		
		if (Deploy.debug) {
		    writer.writeBegin(getIdentifier());
		}

		writeThis(trans, writer);

		if (Deploy.debug) {
		    writer.writeEnd();
		}
		
		container.writeEncrypt(writer, slot.address(), 0);

		if (isActive()) {
		    setStateClean();
		}
	}

    public boolean writeObjectBegin() {
        if (isDirty()) {
            return beginProcessing();
        }
        return false;
    }

    public void writeOwnID(Transaction trans, ByteArrayBuffer writer) {
        write(trans);
        writer.writeInt(getID());
    }
    
    public SlotChangeFactory slotChangeFactory(){
    	return SlotChangeFactory.SYSTEM_OBJECTS;
    }
    
}
